### PR TITLE
[Google Blockly] Add keyboard navigation cursor options to workspace context menu

### PR DESCRIPTION
**Pull request by mikeharv**
As part of the work to [Create (and implement?) plan for making modal function editor keyboard navigable](https://codedotorg.atlassian.net/browse/CT-284), we should make it possible to utilize all three of Blockly's included cursor modes, at least for the purposes of collecting data about which mode(s) users prefer.

### Cursors
_(Descriptions come from Blockly's [demo](https://google.github.io/blockly-samples/plugins/keyboard-navigation/test/index.html) page)_
The cursor controls how the user navigates the blocks, inputs, fields and connections on a workspace. The demo shows three different cursors:

- **Default Cursor:** This cursor uses previous, next, in, and out to navigate through the different parts of a block. See the [developer documentation](https://developers.google.com/blockly/guides/configure/web/keyboard-nav) for more information.
- **Basic Cursor:** Uses pre order traversal to allow users to navigate through everything using only the previous and next command.
- **Line Cursor:** We tried to make this cursor mimic a text editor. Navigating up and down will take the cursor to the next and previous "line" of code. Navigating in and out will move the cursor through all the fields and inputs in that "line" of code.

This PR adds all three cursors to the wrapper. For some reason, the line cursor isn't already available directly from the Blockly instance, but it can be imported from the plugin. This is a bit odd, but it matches how the three cursor types are accessed in their own tests: https://github.com/google/blockly-samples/blob/435f28200f75362afce241922bb233f9fad9a36d/plugins/keyboard-navigation/test/index.js#L62-L71
In any case, we can add them as read-only properties on the wrapper by assigning them directly or using `wrapReadOnlyProperty`. While I was here, I took a minute to re-alphabetize the list.

In addition to this, I wanted to make it possible/somewhat easy for users to try different cursors so I added some new workspace context menu options which are hidden behind an experiment flag. These new options will show the current cursor in use (as a disabled option) and allow for toggling between the other two. Like all context menu options, they aren't super discoverable; our platform never actually directs users to interact with the menu, so this isn't in any way a final UI.

While I was in this space, I got frustrated by the convoluted way in which context menu option weights were mostly hard-coded. I did a quick audit of the standard options that we expect from core Blockly/plugins and added counters that are incremented programmatically to `registerAllContextMenuItems` that ensure that the options show up in the right order. Note: It is okay to use the same weight for multiple options if they have different scopes (block or workspace) as they won't appear together.

Similar to our accessible themes, here we're leveraging local storage to store the user's selected cursor type. This way, their preference will be remembered between levels and page refreshes.

Initial state (keyboard navigation off) - No changes|Keyboard navigation on, experiment disabled - no changes|Keyboard navigation on, experiment enabled, default cursor selected
-|-|-
image|image|image

@samantha-code suggested she would like to bring the cursor options to some keyboard users and ask about their preferences. These new options make it possible to conduct that research so we can hopefully learn which option(s) are most intuitive and preferred.

## Links

Relates to:
- https://codedotorg.atlassian.net/browse/CT-284

## Testing story

Tested that the new options show when and how expected and that they set the new cursor mode without messing anything related up. Tested that the cursor mode is preserved across workspaces and page loads.

Here's a long-ish video that demonstrates its use. I used additional software to monitor key presses on screen.

https://github.com/code-dot-org/code-dot-org/assets/43474485/8226543a-1f06-4459-9457-3d932a5c3d7f

## Follow-up work

At this time, it is not expected that all UI and block fields will be navigable, but that's a gap we'd ideally close before broadcasting this feature more broadly.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/cdoUtils.js [hunk 1]
```diff
@@ -261,6 +261,15 @@ export function getUserTheme(themeOption) {
   return Blockly.themes[localStorage.blocklyTheme] || themeOption || cdoTheme;
 }
 
+/**
+ * Returns a cursor type, based on the presence of an option in the browser's localStorage.
+ * @param {string} type
+ * @returns {string} one of 'default', 'basic', or 'line'
+ */
+export function getUserCursorType() {
+  return localStorage.blocklyCursor || 'default';
```
Comment (ebeastlake) [inline apps/src/blockly/addons/cdoUtils.js]: Am I interpreting correctly that making the cursor type persistent for a given user is a follow-up task?

File: apps/src/blockly/addons/contextMenu.js [hunk 2]
```diff
@@ -148,21 +153,59 @@ const registerKeyboardNavigation = function () {
...
-    weight: 11,
+    weight,
   };
   GoogleBlockly.ContextMenuRegistry.registry.register(keyboardNavigationOption);
 };
 
+/**
+ * Change cursor type for keyboard navigation
+ */
+const registerCursor = function (cursorType, weight) {
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.js]: nit: Could we call this `registerCursorOptions`?

File: apps/src/blockly/addons/contextMenu.js [hunk 3]
```diff
@@ -148,21 +153,59 @@ const registerKeyboardNavigation = function () {
     },
     callback: function (scope) {
       const controller = Blockly.navigationController;
-      scope.workspace.keyboardAccessibilityMode
-        ? controller.disable(scope.workspace)
-        : controller.enable(scope.workspace);
+      if (scope.workspace.keyboardAccessibilityMode) {
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.js]: nit: It might be work creating a couple of helpers here to improve readability, like:
```
const isAccessibilityModeEnabled = (scope) => scope.workspace.keyboardAccessibilityMode;
const isNavigationEnabled = () => experiments.isEnabled(experiments.KEYBOARD_NAVIGATION);
```
But no strong feelings either way.

File: apps/src/blockly/addons/contextMenu.js [hunk 4]
```diff
@@ -148,21 +153,59 @@ const registerKeyboardNavigation = function () {
...
+      }
+    },
+    callback: function (scope) {
+      localStorage.setItem(BLOCKLY_CURSOR, cursorType);
+      Blockly.navigationController.cursorType = cursorType;
+      setNewCursor(cursorType, scope);
+      Blockly.navigationController.navigation.focusWorkspace(scope.workspace);
+    },
+    scopeType: GoogleBlockly.ContextMenuRegistry.ScopeType.WORKSPACE,
+    id: cursorType + 'CursorOption',
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.js]: nit: `${cursorType}CursorOption`

File: apps/src/blockly/addons/contextMenu.js (hunk 2 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.js]: It can be `registerCursorOption` (not plural). However, none of the other functions in this file use that convention. Possibly worth noting, we borrow the current convention from core Blockly: https://github.com/google/blockly/blob/develop/core/contextmenu_items.ts

File: apps/src/blockly/addons/contextMenu.js [hunk 5]
```diff
@@ -148,21 +153,59 @@ const registerKeyboardNavigation = function () {
...
+        return MenuOptionStates.HIDDEN;
+      } else if (isCurrentCursor(cursorType)) {
+        return MenuOptionStates.DISABLED;
+      } else {
+        return MenuOptionStates.ENABLED;
+      }
+    },
+    callback: function (scope) {
+      localStorage.setItem(BLOCKLY_CURSOR, cursorType);
+      Blockly.navigationController.cursorType = cursorType;
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.js]: `setNewCursor` already handles this line, right?

File: apps/src/blockly/addons/contextMenu.js [hunk 6]
```diff
@@ -148,21 +153,59 @@ const registerKeyboardNavigation = function () {
...
+      setNewCursor(cursorType, scope);
+      Blockly.navigationController.navigation.focusWorkspace(scope.workspace);
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.js]: Can you clarify what this line does, and under what conditions we have to refocus?

File: apps/src/blockly/addons/contextMenu.js (hunk 5 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.js]: Ah yes, good catch! Will clean up.

File: apps/src/blockly/addons/contextMenu.js [hunk 7]
```diff
@@ -209,10 +252,11 @@ const themes = [
     label: msg.blocklyTritanopiaTheme(),
   },
 ];
+const cursors = ['default', 'basic', 'line'];
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.js]: nit: Consider moving this to a Blockly-related constants file and making it all caps.
```
const CURSOR_TYPES = [...]
```

File: apps/src/blockly/addons/contextMenu.js (hunk 6 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.js]: Will add a comment. It sets the navigation state to this workspace and moves the cursor to the top block.

File: apps/src/blockly/addons/contextMenu.js [hunk 8]
```diff
@@ -286,22 +331,41 @@ export function registerHelp() {
...
+  registerEditable(nextBlockOptionWeight++);
+  registerShadow(nextBlockOptionWeight++);
+  registerUnshadow(nextBlockOptionWeight);
+
+  // Custom workspace options
+  registerKeyboardNavigation(nextWorkspaceOptionWeight++);
+  registerAllCursors(nextWorkspaceOptionWeight, cursors);
+  nextWorkspaceOptionWeight += cursors.length;
+  registerDarkMode(nextWorkspaceOptionWeight++);
+  registerThemes(nextWorkspaceOptionWeight, themes);
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.js]: Take or leave an alternate organization idea here:

```
// TODO: Add comments on why magic numbers make sense
const INITIAL_BLOCK_OPTION_WEIGHT = 7;
const INITIAL_WORKSPACE_OPTION_WEIGHT = 3;

const registerAllContextMenuItems = function () {
  unregisterDefaultOptions();
  
  let nextBlockOptionWeight = INITIAL_BLOCK_OPTION_WEIGHT;
  registerCustomBlockOptions(nextBlockOptionWeight);

  let nextWorkspaceOptionWeight = INITIAL_WORKSPACE_OPTION_WEIGHT;
  registerCustomWorkspaceOptions(nextWorkspaceOptionWeight, cursors, themes);
};

function registerCustomBlockOptions(startWeight) {
  const blockOptions = [registerHelp, registerDeletable, registerMovable, registerEditable, registerShadow, registerUnshadow];
  blockOptions.forEach(registerOption => {
    registerOption(startWeight);
    startWeight += 1;
  });
}

function registerCustomWorkspaceOptions(startWeight, cursors, themes) {
  const workspaceOptions = [registerKeyboardNavigation, registerDarkMode];
  workspaceOptions.forEach(registerOption => {
    registerOption(startWeight);
    startWeight += 1;
  });

  registerAllCursors(startWeight, cursors);
  startWeight += cursors.length;

  // No explicit increment needed here since registerThemes internally handles the increment
  registerThemes(startWeight, themes);
}
```

Comment (ebeastlake) [reply apps/src/blockly/addons/contextMenu.js]: I'd probably avoid using pre/post-increment operators as an argument because it feels a little weird to me to remember the timing...
```
If used postfix, with operator after operand (for example, x++), the increment operator increments and returns the value before incrementing.
```
But that's just one person's opinion!

**Comment by ebeastlake** [comment]
Mainly nits. I know you've got a major rebase ahead of you, but nice work! 🎉

File: apps/src/blockly/addons/cdoUtils.js (hunk 1 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoUtils.js]: Yes. Since this is just needed for testing, `localStorage` should be sufficient for getting it to work across levels/projects for a single user session. This is consistent with how we remember themes. I would imagine that we'd implement something better for user settings when we create the accessibility panel to manage these options.

File: apps/src/blockly/addons/contextMenu.ts [hunk 9]
```diff
@@ -404,3 +471,35 @@ function unregisterDefaultOptions() {
...
+  // 2: redoWorkspace
+  let nextWeight = 3;
+
+  // Custom workspace options
+  registerKeyboardNavigation(nextWeight++);
+  registerAllCursors(nextWeight, NAVIGATION_CURSORS);
+  nextWeight += NAVIGATION_CURSORS.length;
+  registerDarkMode(nextWeight++);
+  registerThemes(nextWeight, themes);
+}
```
Comment (mikeharv) [inline apps/src/blockly/addons/contextMenu.ts:524]: @ebeastlake I moved things out into two scope-specific functions here (lines 475-505). I'm still using increment operators, but let me know what you think.

File: apps/i18n/common/en_us.json [hunk 10]
```diff
@@ -469,8 +469,6 @@
   "congratsCertificateThanks": "Thanks for submitting!",
   "congratsCertificateContinue": "Now, keep going with our other courses, or see more options to Learn beyond an Hour below.",
   "congratsNextLevelHeading": "Graduate to the next level",
-  "congratsSelfPacedPlTitle": "Explore more self-paced professional learning modules",
-  "congratsSelfPacedPlDescription": "Keep learning with our robust selection of self-paced professional learning modules. Our self-paced learning works in tandem with our curriculum so you can empower yourself and your students at the pace that’s best for you.",
```
Comment (ebeastlake) [inline apps/i18n/common/en_us.json]: It looks like there's still unintended diff here @mikeharv?

File: apps/src/blockly/constants.ts [hunk 11]
```diff
@@ -65,6 +66,7 @@ export const Renderers = {
   DEFAULT: 'cdo_renderer_thrasos',
 };
 
+export const NAVIGATION_CURSORS = ['default', 'basic', 'line'];
```
Comment (ebeastlake) [inline apps/src/blockly/constants.ts]: Why not refer to these as cursor types? (Since they're not cursor instances.)

File: apps/src/blockly/addons/contextMenu.ts [hunk 12]
```diff
@@ -404,3 +474,35 @@ function unregisterDefaultOptions() {
     GoogleBlockly.ContextMenuRegistry.registry.unregister('workspaceDelete');
   } catch (error) {}
 }
+
+function registerCustomBlockOptions() {
```
Comment (ebeastlake) [inline apps/src/blockly/addons/contextMenu.ts:483]: Could you add a general comment about how the weights work/what they are for? (Or maybe JS Docs-style comments by default in Blockly functions would be useful generally, which would explain the weights on the functions that use them?)

**Comment by ebeastlake** [comment]
It looks like there's still some unnecessary diff, but otherwise, this LGTM!

File: apps/src/blockly/constants.ts (hunk 11 above)
Comment (mikeharv) [reply apps/src/blockly/constants.ts]: Sure!

File: apps/src/blockly/addons/contextMenu.ts (hunk 12 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.ts:483]: Can do!

File: apps/i18n/common/en_us.json (hunk 10 above)
Comment (mikeharv) [reply apps/i18n/common/en_us.json]: Oh, hmm. Must be due to merging in Molly’s branch and staging. I’ll make sure it gets cleaned up.

**Review by ebeastlake — APPROVED** [review]

File: apps/src/blockly/addons/contextMenu.ts [hunk 13]
```diff
@@ -264,16 +309,20 @@ const registerTheme = function (name: Themes, label: string, index: number) {
...
 
-function registerThemes(themes: {name: Themes; label: string}[]) {
+function registerThemes(
+  startingWeight: number,
+  themes: {name: Themes; label: string}[]
+) {
   themes.forEach((theme, index) => {
-    registerTheme(theme.name, theme.label, index);
+    const weight = startingWeight + index;
+    registerTheme(theme.name, theme.label, weight);
```
Comment (molly-moen) [inline apps/src/blockly/addons/contextMenu.ts:317]: do you need to update `registerTheme` to take in `weight` rather than `index`?

File: apps/src/blockly/addons/contextMenu.ts [hunk 14]
```diff
@@ -313,22 +362,47 @@ export function registerHelp() {
...
+    .setCursor(Blockly.getNewCursor(type));
+  if (Blockly.getFunctionEditorWorkspace()) {
+    Blockly.getFunctionEditorWorkspace()
+      ?.getMarkerManager()
+      .setCursor(Blockly.getNewCursor(type));
+  }
+  if (oldCurNode) {
+    markerManager.getCursor()?.setCurNode(oldCurNode);
+  }
+  // Set the navigation state to the workspace and moves the cursor to the top block.
```
Comment (molly-moen) [inline apps/src/blockly/addons/contextMenu.ts:391]: for my information: if this moves the cursor to the top block, what does `setCurNode` do?

File: apps/src/blockly/addons/contextMenu.ts [hunk 15]
```diff
@@ -386,21 +460,75 @@ function setAllWorkspacesTheme(
...
+  // sorted in the order listed here.
+  registerHelp(nextWeight++);
+  registerDeletable(nextWeight++);
+  registerMovable(nextWeight++);
+  registerEditable(nextWeight++);
+  registerShadow(nextWeight++);
+  registerUnshadow(nextWeight);
+}
+
+function registerCustomWorkspaceOptions() {
```
Comment (molly-moen) [inline apps/src/blockly/addons/contextMenu.ts:506]: can you add comment(s) explaining the difference between workspace and block options?

File: apps/src/blockly/addons/contextMenu.ts [hunk 16]
```diff
@@ -386,21 +460,75 @@ function setAllWorkspacesTheme(
...
+  let nextWeight = 3;
+
+  // Custom workspace options. We increment the weight for each so they are automatically
+  // sorted in the order listed here. Cursors and Themes have multiple options so we need to adjust
+  // the weight accordingly.
+  registerKeyboardNavigation(nextWeight++);
+  registerAllCursors(nextWeight, NAVIGATION_CURSOR_TYPES);
+  nextWeight += NAVIGATION_CURSOR_TYPES.length;
+  registerDarkMode(nextWeight++);
+  registerThemes(nextWeight, themes);
```
Comment (molly-moen) [inline apps/src/blockly/addons/contextMenu.ts]: should this be `nextWeight++`?

File: apps/src/blockly/addons/functionEditor.ts [hunk 17]
```diff
@@ -125,6 +125,18 @@ export default class FunctionEditor {
 
     const functionEditorTrashcan = new CdoTrashcan(this.editorWorkspace);
     functionEditorTrashcan.init();
+    // Set primary workspace to be active (until a function is shown).
+    Blockly.common.setMainWorkspace(this.primaryWorkspace);
```
Comment (molly-moen) [inline apps/src/blockly/addons/functionEditor.ts:129]: why is this needed now?

File: apps/src/blockly/googleBlocklyWrapper.ts [hunk 18]
```diff
@@ -706,6 +710,18 @@ function initializeBlocklyWrapper(blocklyInstance: GoogleBlocklyInstance) {
 
     blocklyWrapper.navigationController.addWorkspace(workspace);
 
+    blocklyWrapper.getNewCursor = function (type) {
+      switch (type) {
+        case 'basic':
+          return new blocklyWrapper.BasicCursor() as Cursor;
```
Comment (molly-moen) [inline apps/src/blockly/googleBlocklyWrapper.ts]: do we need to cast all of these cursor types? `blocklyWrapper.Cursor()` should be the right type already, right? Does `BasicCursor` extend from `Cursor`?

**Review by molly-moen — COMMENTED** [review]
Mostly questions for my understanding, looks good!

File: apps/src/blockly/addons/contextMenu.ts (hunk 14 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.ts:391]: As far as I understand it, `setCurNode` just sets the location of the marker - the actual blinking SVG.

File: apps/src/blockly/addons/contextMenu.ts (hunk 16 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.ts]: Not needed. The `++` incrementation happens after the value is accessed so that it's ready for the next option to be registered. This is the last option, so we don't need to increment at all. The comment below this line explains that we'd want to count up (by the number of theme options) if we add anything else to the menu.

File: apps/src/blockly/addons/contextMenu.ts (hunk 15 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.ts:506]: Yes, but it really comes down to whether it's an option for when you right-click a block vs right-click the workspace.

File: apps/src/blockly/addons/contextMenu.ts (hunk 16 above)
Comment (ebeastlake) [reply apps/src/blockly/addons/contextMenu.ts]:
```
The ++ incrementation happens after the value is accessed 
```
Molly and I both asked about this, so it might be worth mentioning this in a comment?

File: apps/src/blockly/addons/contextMenu.ts (hunk 14 above)
Comment (molly-moen) [reply apps/src/blockly/addons/contextMenu.ts:391]: Are the marker and the cursor different? What I'm trying to understand is the comment says we move the cursor to the top block, but the marker is wherever it was before?

File: apps/src/blockly/googleBlocklyWrapper.ts (hunk 18 above)
Comment (mikeharv) [reply apps/src/blockly/googleBlocklyWrapper.ts]: It looks like it's happy without the casting. `LineCursor` extends `BasicCursor` which extends `Cursor`. Will remove!

File: apps/src/blockly/addons/functionEditor.ts (hunk 17 above)
Comment (mikeharv) [reply apps/src/blockly/addons/functionEditor.ts:129]: The active ("main") workspace is the one used for keyboard shortcuts, including the one used to turn on keyboard navigation. Without this, if you toggled navigation on immediately upon page load it would actually enable it for the editor workspace since it was the active one. 

We also set the primary workspace to be active when the editor is closed, and do the opposite when we show the editor for a function.

File: apps/src/blockly/addons/contextMenu.ts (hunk 16 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.ts]: Agreed. Will add!

File: apps/src/blockly/addons/contextMenu.ts [hunk 19]
```diff
@@ -523,12 +521,9 @@ function registerCustomWorkspaceOptions() {
   let nextWeight = 3;
 
   // Custom workspace options. We increment the weight for each so they are automatically
-  // sorted in the order listed here. Cursors and Themes have multiple options so we need to adjust
-  // the weight accordingly.
+  // sorted in the order listed here. The ++ incrementation happens after the value is accessed.
   registerKeyboardNavigation(nextWeight++);
-  registerAllCursors(nextWeight, NAVIGATION_CURSOR_TYPES);
-  nextWeight += NAVIGATION_CURSOR_TYPES.length;
```
Comment (molly-moen) [inline apps/src/blockly/addons/contextMenu.ts]: why could this be removed?

Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.ts]: I opted to give each cursor (and theme) option the same menu weight for simplicity.

File: apps/src/blockly/addons/contextMenu.ts (hunk 14 above)
Comment (mikeharv) [reply apps/src/blockly/addons/contextMenu.ts:391]: Oh good question. I just realized we're doing two contradictory things. When we say `setCurNode(oldCurNode)`, we're basically saying that we should keep the cursor on the same node (e.g. stack/block/field/connection) when the user changes from one cursor type to another. However, we're also moving the cursor to the top block after that. I think the latter makes more sense for the purposes of the testing that Sam wants to conduct.

The code for the preservation of the current node comes from the [test playground](https://google.github.io/blockly-samples/plugins/keyboard-navigation/test/index.html) for the plugin: https://github.com/google/blockly-samples/blob/0d09c5d2dfd31d370ada8393cb002a39ef3d028e/plugins/keyboard-navigation/test/index.js#L72-L74

Will remove logic that cared about the old node!
